File: src/main/java/me/zeroeightsix/kami/module/modules/player/AutoEat.java
package me.zeroeightsix.kami.module.modules.player;

import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.client.settings.KeyBinding;
import net.minecraft.init.Items;
import net.minecraft.item.Item;
import net.minecraft.item.ItemFood;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumHand;
import net.minecraft.util.FoodStats;

/**
 * Created by 086 on 8/04/2018.
 * Updated by polymer on 09/03/20
 * Updated by dominikaaaa on 20/03/20
 * Updated by An-En on 24/03/20
 */
@Module.Info(
        name = "AutoEat",
        description = "Automatically eat when hungry",
        category = Module.Category.PLAYER
)
public class AutoEat extends Module {
    private Setting<Integer> foodLevel = register(Settings.integerBuilder("Below Hunger").withValue(15).withMinimum(1).withMaximum(20).build());
    private Setting<Integer> healthLevel = register(Settings.integerBuilder("Below Health").withValue(8).withMinimum(1).withMaximum(20).build());

    private int lastSlot = -1;
    private boolean eating = false;

    private boolean isValid(ItemStack stack, int food) {
        return (
                (passItemCheck(stack.getItem()) && stack.getItem() instanceof ItemFood && (foodLevel.getValue() - food) >= ((ItemFood) stack.getItem()).getHealAmount(stack)) ||
                (passItemCheck(stack.getItem()) && stack.getItem() instanceof ItemFood && (healthLevel.getValue() - (mc.player.getHealth() + mc.player.getAbsorptionAmount()) > 0f))
        );
    }

    private boolean passItemCheck(Item item) {
        if (item == Items.ROTTEN_FLESH) return false;
        if (item == Items.SPIDER_EYE) return false;
        if (item == Items.POISONOUS_POTATO) return false;
        if (item == Items.FISH && new ItemStack(Items.FISH).getItemDamage() == 3) return false;
		return true;
	}
    
    @Override
    public void onUpdate() {
        if (mc.player == null) return;
        if (eating && !mc.player.isHandActive()) {
            if (lastSlot != -1) {
                mc.player.inventory.currentItem = lastSlot;
                lastSlot = -1;
            }
            eating = false;
            KeyBinding.setKeyBindState(mc.gameSettings.keyBindUseItem.getKeyCode(), false);
            return;
        }
        if (eating) return;

        FoodStats stats = mc.player.getFoodStats();
        if (isValid(mc.player.getHeldItemOffhand(), stats.getFoodLevel())) {
            mc.player.setActiveHand(EnumHand.OFF_HAND);
            eating = true;
            KeyBinding.setKeyBindState(mc.gameSettings.keyBindUseItem.getKeyCode(), true);
            mc.playerController.processRightClick(mc.player, mc.world, EnumHand.OFF_HAND);
        } else {
            for (int i = 0; i < 9; i++) {
                if (isValid(mc.player.inventory.getStackInSlot(i), stats.getFoodLevel())) {
                    lastSlot = mc.player.inventory.currentItem;
                    mc.player.inventory.currentItem = i;
                    eating = true;
                    KeyBinding.setKeyBindState(mc.gameSettings.keyBindUseItem.getKeyCode(), true);
                    mc.playerController.processRightClick(mc.player, mc.world, EnumHand.MAIN_HAND);
                    return;
                }
            }
        }
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/player/AutoEat.kt
package me.zeroeightsix.kami.module.modules.player

import me.zeroeightsix.kami.event.SafeClientEvent
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.module.modules.combat.CombatSetting
import me.zeroeightsix.kami.process.PauseProcess.pauseBaritone
import me.zeroeightsix.kami.process.PauseProcess.unpauseBaritone
import me.zeroeightsix.kami.setting.ModuleConfig.setting
import me.zeroeightsix.kami.util.*
import me.zeroeightsix.kami.util.combat.CombatUtils
import me.zeroeightsix.kami.util.items.*
import me.zeroeightsix.kami.util.threads.runSafe
import me.zeroeightsix.kami.util.threads.safeListener
import net.minecraft.client.settings.KeyBinding
import net.minecraft.init.Items
import net.minecraft.inventory.Slot
import net.minecraft.item.ItemBlock
import net.minecraft.item.ItemFood
import net.minecraft.item.ItemStack
import net.minecraft.item.ItemTool
import net.minecraft.util.EnumHand
import net.minecraftforge.fml.common.gameevent.TickEvent

object AutoEat : Module(
    name = "AutoEat",
    description = "Automatically eat when hungry",
    category = Category.PLAYER
) {
    private val belowHunger by setting("BelowHunger", 15, 1..20, 1)
    private val belowHealth by setting("BelowHealth", 10, 1..20, 1)
    private val eatBadFood by setting("EatBadFood", false)
    private val pauseBaritone by setting("PauseBaritone", true)

    private var lastSlot = -1
    var eating = false; private set

    init {
        onDisable {
            stopEating()
        }

        safeListener<TickEvent.ClientTickEvent> {
            if (it.phase != TickEvent.Phase.START || CombatSetting.isActive()) return@safeListener

            if (shouldEat()) {
                if (isValid(player.heldItemOffhand)) {
                    eat(EnumHand.OFF_HAND)
                } else if (swapToFood()) {
                    eat(EnumHand.MAIN_HAND)
                }
            } else if (eating) {
                stopEating()
            }
        }
    }

    private fun SafeClientEvent.shouldEat() =
        player.foodStats.foodLevel < belowHunger
            || CombatUtils.getHealthSmart(player) < belowHealth

    private fun SafeClientEvent.eat(hand: EnumHand) {
        if (pauseBaritone) pauseBaritone()

        KeyBinding.setKeyBindState(mc.gameSettings.keyBindUseItem.keyCode, true)
        playerController.processRightClick(player, world, hand)

        eating = true
    }

    private fun stopEating() {
        unpauseBaritone()

        runSafe {
            if (lastSlot != -1) {
                swapToSlot(lastSlot)
                lastSlot = -1
            }
            KeyBinding.setKeyBindState(mc.gameSettings.keyBindUseItem.keyCode, false)
            playerController.onStoppedUsingItem(player)
        }

        eating = false

    }

    private fun SafeClientEvent.swapToFood(): Boolean {
        if (isValid(player.heldItemMainhand)) return true

        lastSlot = player.inventory.currentItem
        val hasFoodInSlot = swapToItem<ItemFood> { isValid(it) }

        return if (!hasFoodInSlot) {
            lastSlot = -1
            moveFoodToHotbar()
            false
        } else {
            true
        }
    }

    private fun SafeClientEvent.moveFoodToHotbar() {
        val slotFrom = player.storageSlots.firstItem<ItemFood, Slot> {
            isValid(it)
        } ?: return

        moveToHotbar(slotFrom) {
            val item = it.item
            item !is ItemTool && item !is ItemBlock
        }
    }

    private fun isValid(itemStack: ItemStack): Boolean {
        val item = itemStack.item

        return item is ItemFood
            && item != Items.CHORUS_FRUIT
            && (eatBadFood || !isBadFood(itemStack, item))
    }

    private fun isBadFood(itemStack: ItemStack, item: ItemFood) =
        item == Items.ROTTEN_FLESH
            || item == Items.SPIDER_EYE
            || item == Items.POISONOUS_POTATO
            || item == Items.FISH && (itemStack.metadata == 3 || itemStack.metadata == 2) // Puffer fish, Clown fish
}